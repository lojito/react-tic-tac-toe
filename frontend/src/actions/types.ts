import {
  Categories,
  FirstPlayer,
  GameResult,
  ImageName,
  PlayingLevel,
  SquareLocation,
} from "../types";

export type CategoryAction = {
  type: "CHANGE_CATEGORY";
  category: Categories;
};

export type DisabledAction = {
  type: "CHANGE_DISABLED";
  disabled: boolean;
};

export type FirstAction = {
  type: "CHANGE_FIRST";
  first: FirstPlayer;
};

export type UserImageAction = {
  type: "CHANGE_USER_IMAGE";
  userImage: ImageName;
};

export type ComputerImageAction = {
  type: "CHANGE_COMPUTER_IMAGE";
  computerImage: ImageName;
};

export type PlayingLevelAction = {
  type: "CHANGE_PLAYING_LEVEL";
  level: PlayingLevel;
};

export type GameOverAction = {
  type: "CHANGE_GAME_OVER";
  over: boolean;
};

export type ResultAction = {
  type: "CHANGE_GAME_RESULT";
  result: GameResult;
};

export type WinnersSquaresAction = {
  type: "CHANGE_WINNERS_SQUARES";
  winners: SquareLocation[];
};

export type ActionUnknow = {
  type: "UNKNOWN";
};

export type Action =
  | CategoryAction
  | DisabledAction
  | FirstAction
  | UserImageAction
  | ComputerImageAction
  | PlayingLevelAction
  | GameOverAction
  | ResultAction
  | WinnersSquaresAction
  | ActionUnknow;
