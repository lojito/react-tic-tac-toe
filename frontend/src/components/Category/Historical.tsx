import React, { FC, useContext } from "react";
import { DictionaryContext } from "../../contexts/dictionary/DictionaryContext";
import categories from "../../data/json/categories.json";
import { Categories, CategoryId } from "../../types";
import "./Category.scss";

interface Props {
  categoryId: CategoryId;
}

const Historical: FC<Props> = ({ categoryId }) => {
  const { dictionary } = useContext(DictionaryContext);

  return (
    <div className="categories" data-testid="categories">
      {dictionary.CATEGORY}:{" "}
      {dictionary[(categories as Categories[])[categoryId].name]}
    </div>
  );
};

export default Historical;
