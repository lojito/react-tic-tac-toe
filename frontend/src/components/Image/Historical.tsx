import React, { FC, useContext } from "react";
import { DictionaryContext } from "../../contexts/dictionary/DictionaryContext";
import { ImageFolder, ImageName } from "../../types";
import "./Image.scss";

interface Props {
  folder: ImageFolder;
  userImage: ImageName;
  computerImage: ImageName;
}

const Historical: FC<Props> = ({ folder, userImage, computerImage }) => {
  const {
    dictionary: { IMAGE_USER_LABEL, IMAGE_COMPUTER_LABEL },
  } = useContext(DictionaryContext);
  const url = `${process.env.PUBLIC_URL}/images`;

  return (
    <div className="images historical" data-testid="images-historical">
      <div className="user-image">
        <span>{`${IMAGE_USER_LABEL}: `}</span>
        <img
          className="image-token"
          src={`${url}/${folder}/${userImage}.jpg`}
          alt="user"
        />
      </div>

      <div className="computer-image">
        <span>{`${IMAGE_COMPUTER_LABEL}: `}</span>
        <img
          className="image-token"
          src={`${url}/${folder}/${computerImage}.jpg`}
          alt="computer"
        />
      </div>
    </div>
  );
};

export default Historical;
