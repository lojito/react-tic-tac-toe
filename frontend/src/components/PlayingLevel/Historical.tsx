import React, { FC, useContext } from "react";
import { DictionaryContext } from "../../contexts/dictionary/DictionaryContext";
import { PlayingLevel } from "../../types";
import "./PlayingLevel.scss";

interface Props {
  level: PlayingLevel;
}

const Historical: FC<Props> = ({ level }) => {
  const {
    dictionary: {
      PLAYING_LEVEL_LABEL,
      PLAYING_LEVEL_EASY,
      PLAYING_LEVEL_SMART,
    },
  } = useContext(DictionaryContext);

  return (
    <div className="level" data-testid="level">
      <div>
        {PLAYING_LEVEL_LABEL}:{" "}
        {level === PlayingLevel.Easy ? PLAYING_LEVEL_EASY : PLAYING_LEVEL_SMART}
      </div>
    </div>
  );
};

export default Historical;
