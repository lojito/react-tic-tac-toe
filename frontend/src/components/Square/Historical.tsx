import React, { FC, memo } from "react";
import defaultImage from "../../assets/images/default.jpg";
import { SquareBoard } from "../../types";
import "./Square.scss";

const Square: FC<SquareBoard> = memo(
  ({ disabled, id, imagePath, className }) => {
    let url;
    if (imagePath === "default") {
      url = `url(${defaultImage}`;
    } else {
      url = `url(${process.env.PUBLIC_URL}/images/${imagePath}.jpg)`;
    }

    return (
      <button
        className={className}
        disabled
        id={id}
        style={{ backgroundImage: url }}
        aria-label="historical"
      />
    );
  }
);

export default Square;
