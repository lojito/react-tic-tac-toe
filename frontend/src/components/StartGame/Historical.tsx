import React, { FC, useContext } from "react";
import { DictionaryContext } from "../../contexts/dictionary/DictionaryContext";
import { FirstPlayer, Player } from "../../types";
import "./StartGame.scss";

interface Props {
  first: FirstPlayer;
}

const Historical: FC<Props> = ({ first }) => {
  const {
    dictionary: {
      GAME_STARTED_BY,
      GAME_STARTED_BY_COMPUTER,
      GAME_STARTED_BY_YOU,
    },
  } = useContext(DictionaryContext);

  return (
    <div className="start" data-testid="start">
      {GAME_STARTED_BY}:{" "}
      {first === Player.Computer
        ? GAME_STARTED_BY_COMPUTER
        : GAME_STARTED_BY_YOU}
    </div>
  );
};

export default Historical;
