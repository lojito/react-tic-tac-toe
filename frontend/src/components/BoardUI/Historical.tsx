import React, { FC, memo } from "react";
import {
  GameBoard,
  ImageFolder,
  ImageName,
  Player,
  SquareLocation,
} from "../../types";
import Square from "../Square/Historical";
import "./BoardUI.scss";

interface Props {
  folder: ImageFolder;
  computerImage: ImageName;
  userImage: ImageName;
  board: GameBoard;
  winners: SquareLocation[];
}

const Historical: FC<Props> = ({
  folder,
  userImage,
  computerImage,
  board,
  winners,
}) => {
  const boardJSX = board.map((player: Player, square: SquareLocation) => {
    const imagePath =
      player === Player.Computer
        ? `${folder}/${computerImage}`
        : player === Player.User
        ? `${folder}/${userImage}`
        : "default";
    const className = winners.indexOf(square) !== -1 ? "win" : "";

    return (
      <Square
        disabled
        id={square + ""}
        imagePath={imagePath}
        key={`square-${square}`}
        className={className}
      />
    );
  });

  return <div className="board">{boardJSX}</div>;
};

export default memo(Historical);
