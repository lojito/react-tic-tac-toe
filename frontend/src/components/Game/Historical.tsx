import { useMutation } from "@apollo/client";
import React, { FC, useContext, useEffect } from "react";
import { Link, useHistory } from "react-router-dom";
import { DictionaryContext } from "../../contexts/dictionary/DictionaryContext";
import categories from "../../data/json/categories.json";
import DELETE_GAME from "../../mutations/deleteGame";
import GET_GAMES from "../../queries/getGames";
import { Categories, DBGame, GameResult } from "../../types";
import BoardUI from "../BoardUI/Historical";
import Category from "../Category/Historical";
import Images from "../Image/Historical";
import PlayingLevel from "../PlayingLevel/Historical";
import StartGame from "../StartGame/Historical";
import "./Game.scss";

interface Props {
  token: string;
  game: DBGame;
  displayBoard: boolean;
}

const Historical: FC<Props> = ({ game, displayBoard, token }) => {
  const history = useHistory();
  const {
    dictionary: {
      MESSAGE_LOST,
      MESSAGE_WON,
      MESSAGE_DRAW,
      BUTTON_SEE_GAME,
      BUTTON_BACK,
      BUTTON_DELETE_GAME,
      GAME_DELETING,
      GAME_DELETE_ERROR,
    },
  } = useContext(DictionaryContext);

  const context = { headers: { Authorization: `Bearer ${token}` } };
  const [deleteGame, { data, loading, error }] = useMutation(DELETE_GAME, {
    context,
    refetchQueries: [{ query: GET_GAMES, context }],
  });

  const {
    categoryId,
    computerImage,
    userImage,
    first,
    level,
    board,
    winners,
    result,
  } = game;

  const id = game._id!;

  const handleDeleteGame = async (gameId: string) => {
    deleteGame({ variables: { gameId: gameId } });
  };

  useEffect(() => {
    if (data) {
      history.push("/historical");
    }
  }, [data, history]);

  return (
    <div className="game-wrapper">
      <Category categoryId={categoryId} />

      <Images
        userImage={userImage}
        computerImage={computerImage}
        folder={(categories as Categories[])[categoryId].folder}
      />

      <div className="settings">
        <StartGame first={first} />
        <PlayingLevel level={level} />
      </div>

      {displayBoard ? (
        <BoardUI
          userImage={userImage}
          computerImage={computerImage}
          board={board}
          winners={winners}
          folder={(categories as Categories[])[categoryId].folder}
        />
      ) : null}

      <div className="board-footer">
        <span>
          {result === GameResult.Lost
            ? MESSAGE_LOST
            : result === GameResult.Won
            ? MESSAGE_WON
            : MESSAGE_DRAW}
        </span>

        <div className="buttons">
          {!displayBoard ? (
            <button className="see-game">
              <Link to={`/historical/${id}`}>{BUTTON_SEE_GAME}</Link>
            </button>
          ) : null}

          {displayBoard ? (
            <button className="back">
              <Link to="/historical">{BUTTON_BACK}</Link>
            </button>
          ) : null}

          <button
            className="delete-game"
            onClick={() => {
              handleDeleteGame(id);
            }}
          >
            {BUTTON_DELETE_GAME}
          </button>
        </div>

        {loading && <div>{GAME_DELETING}</div>}

        {error && <div className="info">{GAME_DELETE_ERROR}</div>}
      </div>

      {!displayBoard ? <hr /> : null}
    </div>
  );
};

export default Historical;
