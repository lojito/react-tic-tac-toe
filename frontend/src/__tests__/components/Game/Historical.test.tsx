import React from "react";
import { useMutation } from "@apollo/client";
import "@testing-library/jest-dom/extend-expect";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import Historical from "../../../components/Game/Historical";
import { DictionaryContext } from "../../../contexts/dictionary/DictionaryContext";
import { DBGame, GameResult, Player, PlayingLevel } from "../../../types";

jest.mock("../../../components/Category/Historical", () => () => (
  <div>Category: Soccer players</div>
));

jest.mock("../../../components/Image/Historical", () => () => (
  <>
    <div>User: Messi</div>
    <div>Computer: Ronaldo</div>
  </>
));

jest.mock("../../../components/StartGame/Historical", () => () => (
  <div>You start: Yes</div>
));

jest.mock("../../../components/PlayingLevel/Historical", () => () => (
  <div>Playing Level: Easy</div>
));

jest.mock("../../../components/BoardUI/Historical", () => () => (
  <div>BoardUI</div>
));

jest.mock("@apollo/client", () => ({
  useMutation: jest.fn(),
}));

describe("Historical", () => {
  const gameId = "6354074f46bfb3571fe98ebc";

  const mockDBGame: DBGame = {
    _id: gameId,
    board: [2, 2, 2, 0, 1, 1, 1, 0, 1],
    categoryId: 5,
    first: Player.User,
    computerImage: 4,
    userImage: 3,
    level: PlayingLevel.Smart,
    result: GameResult.Won,
    winners: [0, 1, 2],
    createdAt: "2022-10-22T15:07:59.495Z",
    updatedAt: "2022-10-22T15:07:59.495Z",
    __v: 0,
  };

  const mockDictionary = {
    BUTTON_SEE_GAME: "See Game",
    BUTTON_BACK: "Back",
    BUTTON_DELETE_GAME: "Delete Game",
    MESSAGE_WON: "You won!",
    MESSAGE_LOST: "You lost!",
    MESSAGE_DRAW: "It's a draw!",
    GAME_DELETING: "Removing the game data from the database...",
    GAME_DELETE_ERROR:
      "An error ocurred while deleting the game from the database.",
  };

  const mockDeleteGame = jest.fn();

  const mockHistoryPush = jest.fn();

  jest.mock("react-router-dom", () => ({
    useHistory: () => ({
      push: mockHistoryPush,
    }),
  }));

  const history = createMemoryHistory();
  history.push = jest.fn();

  interface HistoricalWithProviderProps {
    result?: GameResult;
    displayBoard?: boolean;
    deleteGame?: jest.Mock;
  }

  const TestComponent = ({
    result = GameResult.Lost,
    displayBoard = false,
  }: HistoricalWithProviderProps = {}) => {
    mockDBGame.result = result;

    return (
      <Router history={history}>
        <DictionaryContext.Provider value={{ dictionary: mockDictionary }}>
          <Historical
            game={mockDBGame}
            displayBoard={displayBoard}
            token="token"
          />
        </DictionaryContext.Provider>
      </Router>
    );
  };

  const renderHistorical = ({
    result = GameResult.Lost,
    displayBoard = false,
    deleteGame = mockDeleteGame,
  }: HistoricalWithProviderProps = {}) => {
    (useMutation as jest.Mock).mockReturnValue([
      deleteGame,
      { data: undefined, loading: false, error: undefined },
    ]);

    const utils = render(
      <TestComponent displayBoard={displayBoard} result={result} />
    );

    const categorySoccer = screen.getByText(/category: soccer players/i);
    const userImage = screen.getByText(/user: Messi/i);
    const computerImage = screen.getByText(/computer: Ronaldo/i);
    const startGameYou = screen.getByText(/you start: yes/i);
    const playingLevelEasy = screen.getByText(/playing Level: easy/i);

    let message;
    if (result === GameResult.Lost) {
      message = screen.getByText(mockDictionary.MESSAGE_LOST);
    } else if (result === GameResult.Draw) {
      message = screen.getByText(mockDictionary.MESSAGE_DRAW);
    } else if (result === GameResult.Won) {
      message = screen.getByText(mockDictionary.MESSAGE_WON);
    } else {
      message = null;
    }

    const buttonDelete = screen.getByRole("button", {
      name: mockDictionary.BUTTON_DELETE_GAME,
    });

    return {
      ...utils,
      mockDictionary,
      categorySoccer,
      userImage,
      computerImage,
      startGameYou,
      playingLevelEasy,
      message,
      buttonDelete,
    };
  };

  it("should render correctly", () => {
    const { asFragment } = renderHistorical();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should display the Category component", () => {
    const { categorySoccer } = renderHistorical();

    expect(categorySoccer).toBeInTheDocument();
  });

  it("should display the Image component for the user", () => {
    const { userImage } = renderHistorical();

    expect(userImage).toBeInTheDocument();
  });

  it("should display the Image component for the computer", () => {
    const { computerImage } = renderHistorical();

    expect(computerImage).toBeInTheDocument();
  });

  it("should display the Start game component", () => {
    const { startGameYou } = renderHistorical();

    expect(startGameYou).toBeInTheDocument();
  });

  it("should display the Playing level component", () => {
    const { playingLevelEasy } = renderHistorical();

    expect(playingLevelEasy).toBeInTheDocument();
  });

  it("should not display the board", () => {
    renderHistorical();

    expect(screen.queryByText("BoardUI")).not.toBeInTheDocument();
  });

  it("should display the board", () => {
    const displayBoard = true;

    renderHistorical({ displayBoard });

    expect(screen.getByText("BoardUI")).toBeInTheDocument();
  });

  it("should display the winning game message", () => {
    const result = GameResult.Won;
    const displayBoard = true;

    const { message: messageWon } = renderHistorical({
      result,
      displayBoard,
    });

    expect(messageWon).toBeInTheDocument();
  });

  it("should display the lost game message", () => {
    const displayBoard = true;

    const { message: messageLost } = renderHistorical({ displayBoard });

    expect(messageLost).toBeInTheDocument();
  });

  it("should display the tie game message", () => {
    const result = GameResult.Draw;
    const displayBoard = true;

    const { message: messageDraw } = renderHistorical({
      result,
      displayBoard,
    });

    expect(messageDraw).toBeInTheDocument();
  });

  it("should display the See Game button if displayBoard is false", () => {
    renderHistorical();

    expect(screen.getByRole("link", { name: "See Game" })).toBeInTheDocument();
  });

  it("should not display the See Game button if displayBoard is true", () => {
    const displayBoard = true;

    renderHistorical({ displayBoard });

    expect(
      screen.queryByRole("link", { name: mockDictionary.BUTTON_SEE_GAME })
    ).not.toBeInTheDocument();
  });

  it("should display the Delete Game button", () => {
    renderHistorical();

    expect(
      screen.getByRole("button", { name: mockDictionary.BUTTON_DELETE_GAME })
    ).toBeInTheDocument();
  });

  it("should display the Back button if displayBoard is true", () => {
    const displayBoard = true;

    renderHistorical({ displayBoard });

    expect(
      screen.getByRole("link", { name: mockDictionary.BUTTON_BACK })
    ).toBeInTheDocument();
  });

  it("should not display the Back button if displayBoard is false", () => {
    renderHistorical();

    expect(
      screen.queryByRole("link", { name: mockDictionary.BUTTON_BACK })
    ).not.toBeInTheDocument();
  });

  it("should route to the historical of the game component when clicking on the See Game button", async () => {
    const user = userEvent.setup();
    renderHistorical();
    const buttonSeeGame = screen.getByRole("link", {
      name: mockDictionary.BUTTON_SEE_GAME,
    });

    await user.click(buttonSeeGame);

    expect(buttonSeeGame).toHaveAttribute("href", `/historical/${gameId}`);
  });

  it("should route to the historical of the games component when clicking on the Back button", async () => {
    const user = userEvent.setup();
    const displayBoard = true;
    renderHistorical({ displayBoard });
    const buttonBack = screen.getByRole("link", {
      name: mockDictionary.BUTTON_BACK,
    });

    await user.click(buttonBack);

    expect(buttonBack).toHaveAttribute("href", `/historical`);
  });

  it("should display the loading message while deleting the game data", async () => {
    const user = userEvent.setup();
    const { buttonDelete, rerender } = renderHistorical();

    await user.click(buttonDelete);

    (useMutation as jest.Mock).mockReturnValueOnce([
      mockDeleteGame,
      { loading: true },
    ]);

    rerender(<TestComponent result={GameResult.Draw} />);

    expect(screen.getByText(mockDictionary.GAME_DELETING)).toBeInTheDocument();
  });

  it("should call the delete game handler when displayBoard is false", async () => {
    const user = userEvent.setup();
    const { buttonDelete } = renderHistorical();

    await user.click(buttonDelete);

    expect(mockDeleteGame).toHaveBeenCalled();
  });

  it("should call the delete game handler when displayBoard is true", async () => {
    const user = userEvent.setup();
    const displayBoard = true;
    const { buttonDelete } = renderHistorical({ displayBoard });

    await user.click(buttonDelete);

    expect(mockDeleteGame).toHaveBeenCalled();
  });

  it("should display an error when trying to delete a game and displayBoard is false", async () => {
    const user = userEvent.setup();
    const { buttonDelete, rerender } = renderHistorical();

    await user.click(buttonDelete);

    (useMutation as jest.Mock).mockReturnValueOnce([
      mockDeleteGame,
      { error: "There was an error saving the game data." },
    ]);

    rerender(<TestComponent result={GameResult.Draw} />);

    expect(
      screen.getByText(mockDictionary.GAME_DELETE_ERROR)
    ).toBeInTheDocument();
  });

  it("should display an error when trying to delete a game and displayBoard is true", async () => {
    const user = userEvent.setup();
    const displayBoard = true;
    const { buttonDelete, rerender } = renderHistorical({ displayBoard });

    await user.click(buttonDelete);

    (useMutation as jest.Mock).mockReturnValueOnce([
      mockDeleteGame,
      { error: "There was an error saving the game data." },
    ]);

    rerender(<TestComponent displayBoard={true} />);

    expect(
      screen.getByText(mockDictionary.GAME_DELETE_ERROR)
    ).toBeInTheDocument();
  });

  it("should route to the historical of the games when deleting a game without error", async () => {
    const user = userEvent.setup();
    const { buttonDelete, rerender } = renderHistorical();

    await user.click(buttonDelete);

    (useMutation as jest.Mock).mockReturnValueOnce([
      mockDeleteGame,
      { data: "the game data" },
    ]);

    rerender(<TestComponent result={GameResult.Draw} />);

    expect(history.push).toHaveBeenCalled();
  });
});
