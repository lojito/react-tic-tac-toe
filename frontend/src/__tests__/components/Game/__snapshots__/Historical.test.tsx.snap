// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Historical should render correctly 1`] = `
<DocumentFragment>
  <div
    class="game-wrapper"
  >
    <div>
      Category: Soccer players
    </div>
    <div>
      User: Messi
    </div>
    <div>
      Computer: Ronaldo
    </div>
    <div
      class="settings"
    >
      <div>
        You start: Yes
      </div>
      <div>
        Playing Level: Easy
      </div>
    </div>
    <div
      class="board-footer"
    >
      <span>
        You lost!
      </span>
      <div
        class="buttons"
      >
        <button
          class="see-game"
        >
          <a
            href="/historical/6354074f46bfb3571fe98ebc"
          >
            See Game
          </a>
        </button>
        <button
          class="delete-game"
        >
          Delete Game
        </button>
      </div>
    </div>
    <hr />
  </div>
</DocumentFragment>
`;
